src/contracts.rs

```diff
@@ -1,70 +1,126 @@
+/// │             │    member      │                     │   IN    │   OUT   │  OUT  │       │
+/// ╞═════════════╪════════════════╪═════════════════════╪═════════╪═════════╪═══════╪═══════╡
+/// │     (1)     │                │                     │         │         │       │       │
+/// ╞═════════════╪════════════════╪═════════════════════╪═════════╪═════════╪═══════╪═══════╡
+/// │ execute     │ execute_fn     │ ContractClosure     │   T1    │    C    │  E1   │   Q   │
+/// ├─────────────┼────────────────┼─────────────────────┼─────────┼─────────┼───────┼───────┤
+/// │ instantiate │ instantiate_fn │ ContractClosure     │   T2    │    C    │  E2   │   Q   │
+/// ├─────────────┼────────────────┼─────────────────────┼─────────┼─────────┼───────┼───────┤
+/// │ query       │ query_fn       │ QueryClosure        │   T3    │ Binary  │  E3   │   Q   │
+/// ├─────────────┼────────────────┼─────────────────────┼─────────┼─────────┼───────┼───────┤
+/// │ sudo        │ sudo_fn        │ PermissionedClosure │   T4    │    C    │  E4   │   Q   │
+/// ├─────────────┼────────────────┼─────────────────────┼─────────┼─────────┼───────┼───────┤
+/// │ reply       │ reply_fn       │ ReplyClosure        │  Reply  │    C    │  E5   │   Q   │
+/// ├─────────────┼────────────────┼─────────────────────┼─────────┼─────────┼───────┼───────┤
+/// │ migrate     │ migrate_fn     │ PermissionedClosure │   T6    │    C    │  E6   │   Q   │
+/// └─────────────┴────────────────┴─────────────────────┴─────────┴─────────┴───────┴───────┘
+/// ```
+/// The general schema depicting which generic type is used in entry points is shown below.
+/// Entry point, when called, is provided minimum two arguments: custom query of type **Q**
+/// (inside `Deps` or `DepsMut`) and input message of type **T1**, **T2**, **T3**, **T4**,
+/// **Reply** or **T6**. As a result, entry point returns custom output message of type
+/// Response<**C**> or **Binary** and an error of type **E1**, **E2**, **E3**, **E4**, **E5**
+/// or **E6**.
+///
+/// ```text
+///    entry_point(query, .., message_in) -> Result<message_out, error>
+///                  ┬           ┬                      ┬          ┬
+///             Q >──┘           │                      │          └──> E1,E2,E3,E4,E5,E6
+///    T1,T2,T3,T4,Reply,T6 >────┘                      └─────────────> C,Binary
+/// ```
+/// Generic type **C** defines a custom message that specific for the **whole blockchain**.
+/// Similarly, the generic type **Q** defines a custom query that is also specific
+/// to the **whole blockchain**. Other generic types a specific to the implemented contract.
+/// So all smart contracts used in the same blockchain will have the same types for **C** and **Q**,
+/// but each contract may use different type for other generic types.
+/// It means that e.g. **T1** in smart contract `A` may differ from **T1** in smart contract `B`.
```

Comment:
Let's add a link to the cosmwasm documentation that describes the custom message types here

Reply:
Custom messages and queries are specific to every blockchain, so it is hard to find a single source of truth.